### DISCUSSION TITLE
homophily_ratio on batches

### DISCUSSION THREAD

**Discussion by RasmusOrsoe** [pyg-team/pytorch_geometric#3028] [Q&A]
Hi!

torch_geometric.utils.homophily_ratio seems to output a single value for a batch of graphs. I'd like to extract this value on a per-graph level, such that instead of a single number, the output would be [batch_size,1].

I realize I could simply calculate this quantity when the graphs are constructed, as a preprocessing step, but for my specific problem the edges change dynamically in the model, and I'd like to be able to calculate this on the fly during training.

I've tried a naive approach in the forward function

```
graphs = data.num_graphs
batch  = data.batch
h_t      = torch.zeros(len(torch.unique(batch)))
for idx in range(0,graphs):
    index = batch == idx
    graph = x[index,:]    
    #edge_index_s = edge_index[:,index]
    h_t[count] = homophily_ratio(edge_index_s, graph[:,3])
```

but I struggle with decomposing the batch-level edge_index into a graph-level edge_index!

Do you have any ideas?

**Comment by wsad1** [comment]
Try this out `edge_index_s = edge_index[:,batch[edge_index[0]]==idx]`.
In general, I think it might be useful to upgrade [homophily](https://pytorch-geometric.readthedocs.io/en/1.7.2/modules/utils.html#torch_geometric.utils.homophily) to support `batch`. I'll try to create a PR with this in a few days.

**Reply by RasmusOrsoe** [reply]
Hi! I tried your suggestion and with a slight modification it ended up working.

```
graphs = data.num_graphs
batch  = data.batch
h_t      = torch.zeros(len(torch.unique(batch)))
for idx in range(0,graphs):    
    edge_index_s = edge_index[:,batch[edge_index[0]]==idx]
    h_t[count] = homophily_ratio(edge_index_s, data.x[batch[edge_index[0]]==idx,3])
```

Although this does decrease my training/inference speed by substantial amount. I look forward to seeing your optimized calculation of this in your PR!

**Reply by rusty1s** [reply]
It should be pretty straightforward to support `batch` in `homophily`, *e.g.*:
```python
row, col = edge_index
scatter_add((y[row] == y[col]).to(torch.long), batch[row], dim=0) / scatter_add(torch.ones_like(row), batch[row], dim=0)
```

**Reply by rusty1s** [reply]
This is now added to master thanks to @wsad1.
